src/passes/used.rs

```diff
@@ -36,95 +40,114 @@ impl Used {
     where
         R: IntoIterator<Item = ExportId>,
     {
-        let mut stack = vec![];
-
-        let mut used = Used {
-            imports: HashSet::new(),
-            tables: HashSet::new(),
-            types: HashSet::new(),
-            funcs: HashSet::new(),
-            globals: HashSet::new(),
-            memories: HashSet::new(),
+        let mut used = Used::default();
+        let mut stack = UsedStack {
+            used: &mut used,
+            stack: Vec::new(),
         };
 
         for r in roots {
             match module.exports.arena[r].item {
-                ExportItem::Function(f) => {
-                    if used.funcs.insert(f) {
-                        stack.push(f);
-                    }
-                }
-                ExportItem::Table(t) => {
-                    used.tables.insert(t);
-                    // TODO: add all functions in an anyfunc table
-                }
+                ExportItem::Function(f) => stack.push_func(f),
+                ExportItem::Table(t) => stack.push_table(t),
                 ExportItem::Memory(m) => {
-                    used.memories.insert(m);
+                    stack.used.memories.insert(m);
                 }
                 ExportItem::Global(g) => {
-                    used.globals.insert(g);
+                    stack.used.globals.insert(g);
                 }
             }
         }
 
-        while let Some(f) = stack.pop() {
-            let func = &module.funcs.arena[f];
-            used.types.insert(func.ty(&module));
+        while let Some(item) = stack.stack.pop() {
+            match item {
+                ItemToVisit::Function(f) => {
+                    let func = &module.funcs.arena[f];
+                    stack.used.types.insert(func.ty(&module));
 
-            match &func.kind {
-                FunctionKind::Local(func) => {
-                    let v = &mut UsedVisitor {
-                        func,
-                        used: &mut used,
-                        stack: &mut stack,
-                    };
-
-                    func.entry_block().visit(v);
+                    match &func.kind {
+                        FunctionKind::Local(func) => {
+                            func.entry_block().visit(&mut UsedVisitor {
+                                func,
+                                stack: &mut stack,
+                            });
+                        }
+                        FunctionKind::Import(i) => {
+                            stack.used.imports.insert(i.import);
+                        }
+                        FunctionKind::Uninitialized(_) => unreachable!(),
+                    }
                 }
-                FunctionKind::Import(i) => {
-                    used.imports.insert(i.import);
+                ItemToVisit::Table(t) => {
```

Comment:
FWIW: we could also have two different stacks for tables and functions if you think that would be more concise than defining a "throw away" enum.

Reply:
Yeah it ended up being a bit smaller, so switched to that